### DISCUSSION TITLE
NXP RT1020 data_copy_xip_relocation missing

### DISCUSSION THREAD

**Discussion by Harija4** [zephyrproject-rtos/zephyr#73717] [Q&A] [answered]
**ERROR:**
c:/zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd.exe: zephyr/kernel/libkernel.a(init.c.obj): in function `z_bss_zero': C:/zephyrproject/zephyr/kernel/init.c:171: undefined reference to `bss_zeroing_relocation' c:/endor/zephyr-sdk-0.15.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.1.0/../../../../arm-zephyr-eabi/bin/ld.bfd.exe: zephyr/kernel/libkernel.a(xip.c.obj): in function `z_data_copy': C:/zephyrproject/zephyr/kernel/xip.c:45: undefined reference to `data_copy_xip_relocation'

I got the above error in my project while running with MIMXRT1020 board file. When I searched for these functions "bss_zeroing_relocation, data_copy_xip_relocation" I couldn't find any function definitions. So I added a dummy function in these respective file (init.c, xip.c). Whether adding a dummy function will create any impact?. If so, then what kind of information should be present inside this function?. 

What is the purpose of these two functions in xip.c and init.c files?.

Could you please help us on this query? @DerekSnell

**Comment by jacob-wienecke-nxp** [comment, marked as answer]
The two functions, `bss_zeroing_relocation`, `data_copy_xip_relocation` are used for to relocating memory to a region other than the default. They are not included unless CONFIG_CODE_DATA_RELOCATION is enabled. The APIs to relocate Code and Data are part of Zephyr's [Code and Data Relocation feature](https://docs.zephyrproject.org/latest/kernel/code-relocation.html).

Based on the output provided, it seems likely to me that you have `CONFIG_CODE_DATA_RELOCATION=y` in your app.

To replicate the issue, I build the hello_world sample application using zephyr v3.6.0. **Without** `CONFIG_CODE_DATA_RELOCATION=y` in the app, the hello_world sample builds without issues. When I add `CONFIG_CODE_DATA_RELOCATION=y` within prj.conf and try to build, I get a linker error. This is because I have not relocated anything.

To ensure that `bss_zeroing_relocation()` is included, I need to relocate some BSS. I added the following line to CMakeLists.txt: `zephyr_code_relocate(FILES src/main.c LOCATION OCRAM_DATA_BSS)`. Following this inclusion, the hello_world application builds again.

I ask that you confirm the zephyr version that you are using as well as share any details from your build that differ from the steps I have performed. It seems that you are on zephyr-sdk-0.15.1, which is older. If needed, you can try using zephyr v3.6 and the recent tools to replicate the working build using my steps.

Thanks, Jacob
